v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dao/spanner/SpannerDao.java

```diff
@@ -82,10 +86,33 @@ public ShadowTableRecord getShadowTableRecord(
     }
   }
 
-  public void updateShadowTable(Mutation mutation) {
-    List<Mutation> mutations = new ArrayList<>();
-    mutations.add(mutation);
-    spannerAccessor.getDatabaseClient().write(mutations);
+  public ShadowTableRecord readShadowTableRecordWithExclusiveLock(
+      String shadowTableName,
+      com.google.cloud.spanner.Key primaryKey,
+      Ddl shadowTableDdl,
+      TransactionContext rwTransaction) {
+    List<String> readColumnList =
+        Arrays.asList(Constants.PROCESSED_COMMIT_TS_COLUMN_NAME, Constants.RECORD_SEQ_COLUMN_NAME);
+    Statement sql =
+        SpannerReadUtils.generateReadSQLWithExclusiveLock(
+            shadowTableName, readColumnList, primaryKey, shadowTableDdl);
+    ResultSet resultSet = rwTransaction.executeQuery(sql);
+
+    // This is the first event for the primary key and hence the latest event.
+    if (!resultSet.next()) {
+      return null;
+    }
+    Struct row = resultSet.getCurrentRowAsStruct();
+    return new ShadowTableRecord(
+        row.getTimestamp(readColumnList.get(0)), row.getLong(readColumnList.get(1)));
+  }
+
+  public DatabaseClient getDatabaseClient() {
+    return spannerAccessor.getDatabaseClient();
+  }
+
+  public void updateShadowTable(Mutation mutation, TransactionContext transactionContext) {
```

Comment:
is there any value in keeping this method ? It uses no class level variables and it pretty basic.